### DISCUSSION TITLE
Typed SQL like query

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#1266] [Ideas]
For a Java developer, it could be quite painful coding a SQL string in the code. It would be cool to have a typed query like this:

```java
database.select().fromType("Vertex")
  .where().property("id").gt().value(100)
  .and().property("name").eq().parameter("p1").vertices()
```

Also cutting off the SQL parser means much faster queries. The query should be strongly typed to avoid syntax errors. Like when you write `database.select().fromType("Vertex").where().property("id").` your next option from the IDE must be the operators.

I have a working branch on my computer I'm using for a while that is under review.

WDYT? Any cool idea to implement before it's out?

Also, do you prefer this long form or a shorter version where `property()` is just `p()`, `value()` just `v()` and `parameter()` just `param()`? The example above would be:

```java
database.select().fromType("Vertex")
  .where().p("id").gt().v(100)
  .and().p("name").eq().param("p1").vertices()
```

**Comment by gramian** [comment]
I really like this!

I am fine with the long version (`property()`, `value()`, `parameter()`), but a medium version (`prop()`, `val()`, `param()`) would also be good, the short version (`p()`, `v()`) is too short for my taste.

**Reply by guilherme-alves-silve** [reply]
Yeah, i like the prop(), val() and param() too. But val can't have incompatibility with lombok, i don't know if it's a issue.

**Comment by guilherme-alves-silve** [comment]
I good referente it's the library JOOQ, i used a lot (https://www.jooq.org/) or even the apache spark. It feels more natural to me the code below:

database.select().fromType("Vertex")
  .where(property("id").gt().value(100)
              .and().property("name").eq().parameter("p1")).vertices()

**Reply by lvca** [reply]
So the condition inside the where method. This could be a good idea to manage parenthesis too.

**Reply by lvca** [reply]
We discussed the syntax and we all prefer avoiding having nested constructs, so all the clauses will be appended to the select object without any nesting.

**Comment by lvca** [comment]
Ok, merged into 23.10.1-SNAPSHOT. It will be released as experimental in 23.10.1 and most likely final on v.23.11.1 next month.
